packages/components/src/edit-in-place-control/index.js

```diff
@@ -0,0 +1,82 @@
+/**
+ * External dependencies
+ */
+import { noop } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { useState, useRef, useEffect } from '@wordpress/element';
+import { ENTER, ESCAPE } from '@wordpress/keycodes';
+
+/**
+ * Internal dependencies
+ */
+import Button from '../button';
+
+function EditInPlaceControl( { label = '', onClick = noop, onUpdate = noop } ) {
+	const [ edit, setEdit ] = useState( false );
+	const [ value, setValue ] = useState( label );
+	const prevValue = useRef();
+
+	const inputRef = useRef();
+	const buttonRef = useRef();
+
+	useEffect( () => {
+		prevValue.current = value;
+		if ( edit ) {
+			inputRef.current.focus();
+			inputRef.current.select();
+		} else {
+			buttonRef.current.focus();
+		}
+	}, [ edit ] );
+
+	return (
+		<>
+			{ ! edit && (
+				<Button
+					ref={ buttonRef }
+					className="components-edit-in-place-control__label"
+					onClick={ () => {
+						setEdit( true );
+						onClick();
+					} }
+				>
+					{ value }
+				</Button>
+			) }
+			{ edit && (
+				<input
+					ref={ inputRef }
+					className="components-edit-in-place-control__input"
+					value={ value }
+					onChange={ ( event ) => {
+						setValue( event.target.value );
+					} }
+					onBlur={ () => {
+						setEdit( false );
+						onUpdate( value );
+					} }
+					onKeyDown={ ( event ) => {
```

Comment:
this is addressed in this PR https://github.com/WordPress/gutenberg/pull/28614

Reply:
This is a very complex thing to get right, which is why I'd rather we don't try to implement this component in this PR.

I'm not completely sure about #28614 as it really breaks the idea of the navigable toolbar—arrow keys are supposed to be for moving between toolbar items. This is also why you're not supposed to put text inputs directly into toolbars.

It doesn't handle `tab` as I mentioned above, but if it did it'd break the Template Part block's toolbar (which is buggy anyway). It also doesn't handle inputs like checkboxes in the expected way, checkboxes don't need to handle arrow key navigation.

We're really just running into all the same problems as I had on https://github.com/WordPress/gutenberg/pull/25890. The accessibility team's solution is to move this into a popover, which is impossible to do nicely with the given design.

Also related is the console warning on this PR:
```
Using custom components as toolbar controls is deprecated. Please use ToolbarItem or ToolbarButton components instead. See: https://developer.wordpress.org/block-editor/components/toolbar-button/#inside-blockcontrols
```

Both the input and button item need to be implemented using those components, but doing so would break how the toolbar works, the input shouldn't have a roving tab index.

Reply:
I commented on #28614 with more context. But is it possible to put this control in a popover? Otherwise, we should maybe consider this is not really a toolbar, but just a local header with tabbable elements.